src/OrleansAzureUtils/MultiClusterNetwork/GossipTableInstanceManager.cs

```diff
@@ -0,0 +1,342 @@
+
+        public static string ConstructRowKey(SiloAddress silo, string clusterid)
+        {
+            return String.Format("{1}{0}{2}{0}{3}{0}{4}", SEPARATOR, clusterid, silo.Endpoint.Address, silo.Endpoint.Port, silo.Generation);
+        }
+
+
+
+        internal void UnpackRowKey()
+        {
+            var debugInfo = "UnpackRowKey";
+            try
+            {
+                int idx3 = RowKey.LastIndexOf(SEPARATOR);
+                int idx2 = RowKey.LastIndexOf(SEPARATOR, idx3 - 1);
+                int idx1 = RowKey.LastIndexOf(SEPARATOR, idx2 - 1);
+
+                ClusterId = RowKey.Substring(0, idx1);
+                var addressstr = RowKey.Substring(idx1 + 1, idx2 - idx1 - 1);
+                var portstr = RowKey.Substring(idx2 + 1, idx3 - idx2 - 1);
+                var genstr = RowKey.Substring(idx3 + 1);
+                Address = IPAddress.Parse(addressstr);
+                Port = Int32.Parse(portstr);
+                Generation = Int32.Parse(genstr);
+
+                this.SiloAddress = SiloAddress.New(new IPEndPoint(Address, Port), Generation);
+            }
+            catch (Exception exc)
+            {
+                throw new AggregateException("Error from " + debugInfo, exc);
+            }
+        }
+
+        internal MultiClusterConfiguration ToConfiguration()
+        {
+            string clusterliststring = Clusters;
+            var clusterlist = clusterliststring.Split(',');
+            return new MultiClusterConfiguration(GossipTimestamp, clusterlist, Comment ?? "");
+        }
+
+        internal GatewayEntry ToGatewayEntry()
+        {
+            // call this only after already unpacking row key
+            return new GatewayEntry()
+            {
+                ClusterId = ClusterId,
+                SiloAddress = SiloAddress,
+                Status = (GatewayStatus) Enum.Parse(typeof(GatewayStatus), Status),
+                HeartbeatTimestamp = GossipTimestamp
+            };
+        }
+
+        public override string ToString()
+        {
+            if (RowKey == CONFIGURATION_ROW)
+                return ToConfiguration().ToString();
+            else
+                return string.Format("{0} {1}",
+                    this.SiloAddress, this.Status);
+        }
+    }
+
+    internal class GossipTableInstanceManager
+    {
+        public string TableName { get { return INSTANCE_TABLE_NAME; } }
+
+        private const string INSTANCE_TABLE_NAME = "OrleansGossipTable";
+
+        private readonly AzureTableDataManager<GossipTableEntry> storage;
+        private readonly TraceLogger logger;
+
+        internal static TimeSpan initTimeout = AzureTableDefaultPolicies.TableCreationTimeout;
+
+        public string GlobalServiceId { get; private set; }
+
+        private GossipTableInstanceManager(string globalServiceId, string storageConnectionString, TraceLogger logger)
+        {
+            GlobalServiceId = globalServiceId;
```

Comment:
since this is going to be used as the PartitionKey, you might want to sanitize it (`AzureStorageUtils.SanitizeTableProperty(key)`)